### ISSUE TITLE
Laplace and exponential noise distributions

### ISSUE THREAD

**Issue by pepper-jk** [pytorch/opacus#59]
##  Feature

Allowing other distributions for the noise created by the `privacy_engine`.
Especially Laplace and exponential.

The current implementation of the `privacy_engine` allows only the use of the normal distribution to create noise.

## Alternatives

The `privacy_engine` uses `torch.normal`, but there is no such functions as `torch.laplace` or `torch.exponential`.
This is why I would suggest using the `torch.distributions` distributions.
This also appears to be more current approach according to [this comment](https://github.com/pytorch/pytorch/issues/4620#issuecomment-359732513).

However I assume the reason you still use `torch.normal` is because `torch.distributions.normal.Normal` does not allow a custom `generator` and therefore does not support `urandom`.
So maybe this also requires an adjustment of the `torch.distributions` distributions?

## Additional context

I created a sample implementation of a `privacy_engine_xl` with a modified `_generate_noise` method for our use case.
However I'd prefer the use of `urandom` and also a way to pass a `device` to the distributions.
It seems like `Normal` gets the device from the `loc` (aka `mean`) param if it is a Tensor, however that seems more like a hack than a solution, as the mean used in the `privacy_engine` is `0`.

```python
def _generate_noise(self, max_norm, parameter):
    if self.noise_multiplier > 0:
        mean = 0
        scale = self.noise_multiplier * max_norm

        if self.noise_type == "gaussian":
            dist = torch.distributions.normal.Normal(mean, scale)
        elif self.noise_type == "laplacian":
            dist = torch.distributions.laplace.Laplace(mean, scale)
        elif self.noise_type == "exponential":
            dist = torch.distributions.exponential.Exponential(1 / scale)
        else:
            dist = torch.distributions.normal.Normal(mean, scale)

        noise = dist.sample(parameter.grad.shape)

        return noise
    return 0.0
```

I'd be happy to submit a pull request later on, however I think this can not be the final solution.

Documentation:
https://pytorch.org/docs/master/distributions.html#normal
https://pytorch.org/docs/master/distributions.html#laplace
https://pytorch.org/docs/master/distributions.html#exponential

**Comment by pepper-jk** [comment]
At least for [exponential](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.exponential_) and [normal](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.normal_) distribution there appears to be an alternative approach.

One can create a Tensor and fill it with numbers from the given distribution **in-place**, as described [here](https://pytorch.org/docs/stable/torch.html#in-place-random-sampling).
The in-place sampling even allows a generator, so the `secure_generator` (now called `self.random_number_generator`) created in the `privacy_engine` could be utilized and therefore `urandom` could be used as a source.
The device can be given to a tensor, when initiating a tensor object.

**However this approach lacks Laplace, again.**
It is really annoying that the distributions are not standardized so that they behave the same (regarding parameters and configurations) no matter where we call them in the `torch` libs.

**Comment by pepper-jk** [comment]
I updated our `_generate_noise` method to use a tensor for `scale` and `rate` to pass on the device.

So the only thing missing is the generator support. Hoping that our proposal in pytorch/pytorch#45115 gets addressed and we will be able to add it later and therefore use `urandom` as a source.

For our research purposes however this should be sufficient.

```python
def _generate_noise(self, max_norm, parameter):
    if self.noise_multiplier > 0:
        mean = 0
        scale_scalar = self.noise_multiplier * max_norm

        scale = torch.full(size=parameter.grad.shape, fill_value=scale_scalar, dtype=torch.float32, device=self.device)

        if self.noise_type == "gaussian":
            dist = torch.distributions.normal.Normal(mean, scale)
        elif self.noise_type == "laplacian":
            dist = torch.distributions.laplace.Laplace(mean, scale)
        elif self.noise_type == "exponential":
            rate = 1 / scale
            dist = torch.distributions.exponential.Exponential(rate)
        else:
            dist = torch.distributions.normal.Normal(mean, scale)

        noise = dist.sample()

        return noise
    return 0.0
```

**Comment by Darktex** [comment]
Hi @pepper-jk !

It looks like this is blocked on the PyTorch side, but there are a few comments I'd like to add: in general, while I agree that it would be very simple for us to extend support for more distributions, we are hesitant to do that because we don't see a use case where you would want to use an exponential or Laplace distribution for DP-SGD.

That being said, you should absolutely be able to plug in whatever RNG you want without you having to edit our lib or upstream a PR to us necessarily, just like you don't need to commit a new `nn.Module` you wrote to PyTorch every time. We are sorry this is not yet very doable, but we are working on making this lib a lot more modular for research for this exact reason :) 

For now, I would recommend either editing in-place or monkeypatching in your script:

```python
privacy_engine = PrivacyEngine(...)
privacy_engine._generate_noise = my_generate_noise
```

**Comment by pepper-jk** [comment]
Just for the record, We since published a small [library](https://github.com/tklab-tud/aDPtorch) we extended to create exponantial and laplace noise using opacus.
In case anyone else needs this.

We are using it in the code for our [Label Leakage from Gradients paper](https://github.com/tklab-tud/LLG).
